### PR TITLE
fix memory leak during altering distributed table with a lot of partition and shards

### PR DESCRIPTION
**Pull request by aykut-bozkurt**
2 improvements to prevent memory leaks during altering or undistributing distributed tables with a lot of partitions and shards:

1. Free memory for each colocated table (`TableConversionReturn *colocatedReturn = con->function(&cascadeParam)` calls `AlterDistributedTable`, `UndistributeTable`, or `AlterTableSetAccessMethod` for each colocated, which leads an increase in memory usage),
2. Free memory for each partition table (`TableConversionReturn *partitionReturn = con->function(&partitionParam)` calls `AlterDistributedTable`, `UndistributeTable`, or `AlterTableSetAccessMethod` for each partition table, which leads an increase in memory usage),

DESCRIPTION: Fixes a memory leak issue during altering distributed table with a lot of partition and shards.

Fixes https://github.com/citusdata/citus/issues/6503.

### CODE REVIEW

File: src/backend/distributed/commands/alter_table.c [hunk 1]
```diff
@@ -402,7 +415,15 @@ UndistributeTable(TableConversionParameters *params)
 	params->conversionType = UNDISTRIBUTE_TABLE;
 	params->shardCountIsNull = true;
 	TableConversionState *con = CreateTableConversion(params);
-	return ConvertTable(con);
+	TableConversionReturn *tableConversionReturn = ConvertTable(con);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/alter_table.c:418]: I thought accumulating `GenerateAlterTableAttachPartitionCommand` is the problem inside `ConvertTable`.

Is this solving that problem?

File: src/backend/distributed/commands/alter_table.c [hunk 2]
```diff
@@ -441,7 +472,16 @@ AlterDistributedTable(TableConversionParameters *params)
 		ereport(DEBUG1, (errmsg("setting multi shard modify mode to sequential")));
 		SetLocalMultiShardModifyModeToSequential();
 	}
-	return ConvertTable(con);
+
+	TableConversionReturn *tableConversionReturn = ConvertTable(con);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/alter_table.c:476]: I think the main problem is `ConvertTable`, so we should only take that into a mem. context, not the whole AlterDistributedTable/AlterTableSetAccessMethod/UndistributeTable etc.

Instead of copy & pasting the same pattern to all over, can we create a single function:
```python

def ConvertTable:
         create mem context
         return = ConvertTableInternal (rename the current ConvertTable to ConvertTableInternal)
         switch to original context
         copyConversion
         delete mem context
         return copyConversion
```

that could make this PR easier to review & merge

**Review by onderkalaci — COMMENTED** [review]
Can we add test for this one?

An example I can think of here: https://github.com/citusdata/citus/blob/main/src/test/regress/expected/columnar_memory.out

Can we have something like that? Today, we also have Postgres' view: https://www.postgresql.org/docs/current/view-pg-backend-memory-contexts.html

It'd be nice to find a way to add a test that ensures the memory is NOT increasing

File: src/backend/distributed/commands/alter_table.c (hunk 2 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:476]: We call AlterDistributedTable, UndistributeTable for each partition and colocated table during alteration of a table. I also wanted to make sure we are freeing memory allocated there. We allocate memory there for foreign key dependecy check.
But, let me check if we can leave it as it is by moving memory context creation-deletion to ConvertTable.

File: src/backend/distributed/commands/alter_table.c (hunk 1 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:418]: We should accumulate only attachPartitionCommands and foreignKeyCommands now. We should refactor the alter table if we want to free those as well. 

I could not reproduce the issue linked with the PR by only leaking above mentioned commands.

File: src/backend/distributed/commands/alter_table.c (hunk 2 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:476]: I think that having a context for AlterDistributedTable, UndistributeTable is a safer approach even if we should repeat the pattern for both. If we also should handle partition related allocations, we can hanldle them in a nested partition context as always. 
e.g.
```c
MemoryContext citusPartitionContext =
			AllocSetContextCreate(CurrentMemoryContext,
								  "citus_per_partition_context",
								  ALLOCSET_DEFAULT_SIZES);
```

File: src/backend/distributed/commands/alter_table.c (hunk 1 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:418]: The main problem is actually where we deparse and execute commands using our utility executor method. We do not free at `ProcessUtilityParseTree` which causes huge chunk of memory to leak when we try to attach all partitions.

Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:418]: Currently we only leak attachPartitionCommands and foreignKeyCommands, but if we still want to free them for each partition, then we should refactor the alter distributed table udf.

I could not reproduce the memory issue linked with the PR by only leaking above mentioned commands. It seems fine to accumulate those.

File: src/backend/distributed/commands/alter_table.c (hunk 2 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c:476]: I refactored it such that we free context only at ConvertTable. Memory allocations at AlterDistributedTable, UndistributeTable functions itself seems negligible.

File: src/backend/distributed/commands/alter_table.c [hunk 3]
```diff
@@ -854,8 +874,12 @@ ConvertTable(TableConversionState *con)
 				.suppressNoticeMessages = con->suppressNoticeMessages
 			};
 			TableConversionReturn *colocatedReturn = con->function(&cascadeParam);
+
+			/* persist foreign key commands to older context */
+			MemoryContextSwitchTo(oldContext);
```
Comment (JelteF) [inline src/backend/distributed/commands/alter_table.c:879]: This is a no-op because the same is already done on line 845. Should line 845 be
```diff
-MemoryContextSwitchTo(oldContext);
-MemoryContextSwitchTo(convertTableContext);
```

(line 901 is also a possible no-op because of the same reason)

File: src/backend/distributed/commands/alter_table.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/alter_table.c]: > I refactored it such that we free context only at ConvertTable

The current code is slightly complex. We create/switch many contexts. Why not [this](https://github.com/citusdata/citus/pull/6726#discussion_r1116857088)?

File: src/backend/distributed/commands/alter_table.c [hunk 4]
```diff
@@ -1315,6 +1346,8 @@ ErrorIfUnsupportedCascadeObjects(Oid relationId)
 	bool unsupportedObjectInDepGraph =
 		DoesCascadeDropUnsupportedObject(RelationRelationId, relationId, nodeMap);
 
+	hash_delete_all(nodeMap);
```
Comment (onderkalaci) [inline src/backend/distributed/commands/alter_table.c:1349]: unrelated change. Why do we need this?

File: src/backend/distributed/commands/foreign_constraint.c [hunk 5]
```diff
@@ -949,7 +949,11 @@ TableReferenced(Oid relationId)
 	int flags = INCLUDE_REFERENCED_CONSTRAINTS | INCLUDE_ALL_TABLE_TYPES;
 	List *foreignKeyOids = GetForeignKeyOids(relationId, flags);
 
-	return list_length(foreignKeyOids) > 0;
+	bool tableReferenced = list_length(foreignKeyOids) > 0;
```
Comment (onderkalaci) [inline src/backend/distributed/commands/foreign_constraint.c:952]: these foreign key related changes should happen in a separate commit

File: src/backend/distributed/commands/alter_table.c (hunk 2 above)
Comment (onderkalaci) [reply src/backend/distributed/commands/alter_table.c]: (as Jelte's comment below shows, this is error prone style)

File: src/backend/distributed/commands/alter_table.c (hunk 4 above)
Comment (aykut-bozkurt) [reply src/backend/distributed/commands/alter_table.c]: This one and also foreign key related one was left over from my local tests. They are unrelated to leakage so I removed them.

**Review by onderkalaci — APPROVED** [review]
Looks good with some minor comments. There is still some memory growth happening, but this PR solves a bigger problem. Let's look further in separate PR(s).

I'd probably backport this to 11.2 (at least)
